model/src/application/interaction/mod.rs

```diff
@@ -1,142 +1,177 @@
 //! Used when receiving interactions through gateway or webhooks.
+//!
+//! See [Discord Docs/Receiving and Responding].
+//!
+//! [Discord Docs/Receiving and Responding]: https://discord.com/developers/docs/interactions/receiving-and-responding
 
 pub mod application_command;
 pub mod message_component;
 pub mod modal;
 
 mod interaction_type;
-mod ping;
 
-use self::modal::ModalSubmitInteraction;
-pub use self::{
-    application_command::ApplicationCommand, interaction_type::InteractionType,
-    message_component::MessageComponentInteraction, ping::Ping,
+use std::fmt::{Formatter, Result as FmtResult};
+
+pub use self::interaction_type::InteractionType;
+use self::{
+    application_command::CommandData, message_component::MessageComponentInteractionData,
+    modal::ModalInteractionData,
 };
 
 use crate::{
     channel::Message,
-    guild::PartialMember,
+    guild::{PartialMember, Permissions},
     id::{
         marker::{ApplicationMarker, ChannelMarker, GuildMarker, InteractionMarker, UserMarker},
         Id,
     },
     user::User,
 };
 use serde::{
-    de::{Deserializer, Error as DeError, IgnoredAny, MapAccess, Visitor},
-    Deserialize, Serialize,
+    de::{Error as DeError, IgnoredAny, MapAccess, Visitor},
+    Deserialize, Deserializer, Serialize,
 };
 use serde_value::{DeserializerError, Value};
-use std::fmt::{Formatter, Result as FmtResult};
 
 /// Payload received when a user executes an interaction.
 ///
-/// Each variant corresponds to `InteractionType` in the Discord Docs. See
-/// [Discord Docs/Interaction Object].
+/// See [Discord Docs/Interaction Object].
 ///
 /// [Discord Docs/Interaction Object]: https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object-interaction-structure
 #[derive(Clone, Debug, Eq, PartialEq, Serialize)]
-#[serde(untagged)]
-#[non_exhaustive]
-pub enum Interaction {
-    /// Ping variant.
-    Ping(Box<Ping>),
-    /// Application command variant.
-    ApplicationCommand(Box<ApplicationCommand>),
-    /// Application command autocomplete variant.
-    ApplicationCommandAutocomplete(Box<ApplicationCommand>),
-    /// Message component variant.
-    MessageComponent(Box<MessageComponentInteraction>),
-    /// Modal submit variant.
-    ModalSubmit(Box<ModalSubmitInteraction>),
+pub struct Interaction {
+    /// App's permissions in the channel the interaction was sent from.
+    ///
+    /// This field is present when the interaction is invoked in a guild.
```

Comment:
(for all relevant fields)
```suggestion
    /// Present when the interaction is invoked in a guild.
```